apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,483 @@
+   * Set profile connect state.
+   *
+   * @private
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      if (profile === 'opp' && this.transferIcon) {
+        this.transferIcon.update();
+      }
+      if (profile === 'a2dp' && this.headphoneIcon) {
+        this.headphoneIcon.update();
+      }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+
+    // hold the instances so we can remove it when needed
+    this._bindBtAdapterHandler =
+      this._btAdapterHandler.bind(this);
+
+    this._bindHfpStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.HFP, evt.status);
+    }.bind(this);
+    this._bindA2dpStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.A2DP, evt.status);
+    }.bind(this);
+    this._bindScoStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.SCO, evt.status);
+    }.bind(this);
+
+    // listen states when the adapter is available
+    window.addEventListener('bluetooth-available',
+      this._adapterAvailableHandler.bind(this));
+
+    this._initDefaultAdapter();
+
+    this._bluetooth.addEventListener('attributechanged',
+      this._btManagerHandler.bind(this));
+
+    // decouple enable/disable bluetooth function from other system part
+    window.addEventListener('request-enable-bluetooth',
+      this._requestEnableHandler.bind(this));
+    window.addEventListener('request-disable-bluetooth',
+      this._requestDisableHandler.bind(this));
+
+    // relay bluetooth transfer system messages
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-start',
+      this._oppTransferStartHandler.bind(this));
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-complete',
+      this._oppTransferCompleteHandler.bind(this));
+
+    // expose isEnabled function to Service.query
+    Service.registerState('isEnabled', this);
```

Comment:
It would be helpful reporting all the states of the current adapter. There are four states reported from gecko: "enabled", "disabled", "enabling", "disabling". These four states can be used to determine the state of the bluetooth toggle in quick settings. Therefore, I would suggest to expose the state and fire the state change events. We can still fire `bluetooth-enabled` and `bluetooth-disabled` but only when the adapter is in the stable states.